logsmanagement/db_api.c

```diff
@@ -0,0 +1,1096 @@
+		 * storage). This is useful if BLOB_MAX_FILES is reduced after an agent
+		 * restart (for example, if in the future it is not hardcoded, but 
+		 * instead it is read from the configuration file). LOGS_TABLE entries
+		 * should be deleted automatically (due to ON DELETE CASCADE). */
+		{
+			sqlite3_stmt *stmt_get_BLOBS_TABLE_size;
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
+				"SELECT MAX(Id) FROM " BLOBS_TABLE ";",
+				-1, &stmt_get_BLOBS_TABLE_size, NULL);
+			if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+			rc = sqlite3_step(stmt_get_BLOBS_TABLE_size);
+			if (unlikely(rc != SQLITE_ROW)) fatal_sqlite3_err(rc, __LINE__);
+			const int blobs_table_max_id = sqlite3_column_int(stmt_get_BLOBS_TABLE_size, 0);
+
+			sqlite3_stmt *stmt_retrieve_filename_last_digits; // This statement retrieves the last digit(s) from the Filename column of BLOBS_TABLE
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
+				"WITH split(word, str) AS ( SELECT '', (SELECT Filename FROM " BLOBS_TABLE " WHERE Id = ? ) || '.' "
+				"UNION ALL SELECT substr(str, 0, instr(str, '.')), substr(str, instr(str, '.')+1) FROM split WHERE str!='' ) "
+				"SELECT word FROM split WHERE word!='' ORDER BY LENGTH(str) LIMIT 1;",
+				-1, &stmt_retrieve_filename_last_digits, NULL);
+			if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+			sqlite3_stmt *stmt_delete_row_by_id; 
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
+				"DELETE FROM " BLOBS_TABLE " WHERE Id = ?;",
+				-1, &stmt_delete_row_by_id, NULL);
+			if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+			for (int id = 1; id <= blobs_table_max_id; id++){
+
+				rc = sqlite3_bind_int(stmt_retrieve_filename_last_digits, 1, id);
+				if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+				rc = sqlite3_step(stmt_retrieve_filename_last_digits);
+				if (unlikely(rc != SQLITE_ROW)) fatal_sqlite3_err(rc, __LINE__);
+				int last_digits = sqlite3_column_int(stmt_retrieve_filename_last_digits, 0);
+				rc = sqlite3_reset(stmt_retrieve_filename_last_digits);
+				if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+				/* If last_digits > BLOB_MAX_FILES - 1, then some BLOB files will need to be removed
+				 * (both from DB BLOBS_TABLE and also from the disk) */
+				if(last_digits > BLOB_MAX_FILES - 1){
+
+					// Delete entry from DB BLOBS_TABLE
+					rc = sqlite3_bind_int(stmt_delete_row_by_id, 1, id);
+					if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+					rc = sqlite3_step(stmt_delete_row_by_id);
+					if (unlikely(rc != SQLITE_DONE)) fatal_sqlite3_err(rc, __LINE__);
+					rc = sqlite3_reset(stmt_delete_row_by_id);
+					if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+					// Delete BLOB file from filesystem
+					char blob_delete_path[FILENAME_MAX + 1];
+					sprintf(blob_delete_path, "%s" BLOB_STORE_FILENAME ".%d", p_file_infos_arr->data[i]->db_dir, last_digits);
```

Comment:
It is the last time that I will write about this, but here we have a clear scenario where it is better to use `snprintfz` instead `snprintf`.

Please, take a look in our Costas' `z` functions (This is @underhood name, that  I agree ), because it makes more sense to use them in this code.

Reply:
The problem is that in certain cases, `snprintfz` is not the same as `snprintf`. Also, I intend to keep using `sprintf` in cases where there is no overflow risk.